File: src/generate.js
import { CREDIT_URL, SPEC_URL, SPEC_VERSION } from "./constants.js";
import { redactSecrets } from "./secrets.js";

const MAX_ACTIONS = 8;

export function generateSkillsMarkdown(scan) {
  const lines = [
    "# Skills",
    "",
    `This file tells agents how to use ${scan.name}.`,
    "",
    "## Purpose",
    "",
    sentence(scan.purpose),
    "",
    "## API",
    "",
    `- Base path: \`${scan.apiBase || "/"}\``,
    `- Source: ${sourceText(scan)}`,
    "",
    "## Auth",
    "",
    `- ${sentence(scan.auth)}`,
    "- Do not put secrets, tokens, or private keys in this file.",
    "",
    "## Safe Actions",
    "",
    ...listOrDefault(scan.safeActions, "Use read-only calls when auth rules allow them."),
    "",
    "## Risky Actions",
    "",
    ...listOrDefault(scan.riskyActions, "Ask before write, delete, email, payment, or admin work."),
    "",
    "## Limits",
    "",
    `- ${sentence(scan.limits)}`,
    "- Stop after repeated errors.",
    "",
    "## More Info",
    "",
    ...moreInfo(scan),
    "",
    "## Credits",
    "",
    `- Spec: ${SPEC_URL}`,
    `- Spec version: ${SPEC_VERSION}`,
    `- Credit: ${CREDIT_URL}`,
    "- License: CC-BY-4.0"
  ];

  return `${redactSecrets(lines.join("\n")).replace(/\n{3,}/g, "\n\n").trim()}\n`;
}

function sourceText(scan) {
  const languages = scan.languages?.length ? scan.languages.join(", ") : "project files";
  const routes = scan.routes?.length ?? 0;
  if (routes === 0) return "No public HTTP routes found";
  if (routes === 1) return `1 route from ${languages}`;
  return `${routes} routes from ${languages}`;
}

function listOrDefault(items, fallback) {
  const clean = [...new Set(items ?? [])]
    .filter(Boolean)
    .slice(0, MAX_ACTIONS)
    .map((item) => `- ${sentence(item)}`);

  if (clean.length === 0) return [`- ${fallback}`];
  return clean;
}

function moreInfo(scan) {
  const lines = [];
  for (const doc of (scan.docs ?? []).slice(0, 6)) {
    lines.push(`- Docs: ${doc}`);
  }
  if (scan.support) lines.push(`- Support: ${scan.support}`);
  if (scan.notes?.length) {
    for (const note of scan.notes.slice(0, 4)) lines.push(`- Note: ${sentence(note)}`);
  }
  if (lines.length === 0) lines.push("- Add docs in `autoskill.config.json`.");
  return lines;
}

function sentence(value) {
  const clean = String(value ?? "")
    .replace(/\s+/g, " ")
    .trim();
  if (!clean) return "";
  if (/[.!?`]$/.test(clean)) return clean;
  return `${clean}.`;
}

File: src/generate.rs
use crate::constants::{CREDIT_URL, SPEC_URL, SPEC_VERSION};
use crate::scan::ScanResult;
use crate::secrets::redact_secrets;

const MAX_ACTIONS: usize = 8;

pub fn generate_skills_markdown(scan: &ScanResult) -> String {
    let mut lines = vec![
        "# Skills".to_string(),
        String::new(),
        format!("This file tells agents how to use {}.", scan.name),
        String::new(),
        "## Purpose".to_string(),
        String::new(),
        sentence(&scan.purpose),
        String::new(),
        "## API".to_string(),
        String::new(),
        format!("- Base path: `{}`", empty_default(&scan.api_base, "/")),
        format!("- Source: {}", source_text(scan)),
        String::new(),
        "## Auth".to_string(),
        String::new(),
        format!("- {}", sentence(&scan.auth)),
        "- Do not put secrets, tokens, or private keys in this file.".to_string(),
        String::new(),
        "## Safe Actions".to_string(),
        String::new(),
    ];
    lines.extend(list_or_default(
        &scan.safe_actions,
        "Use read-only calls when auth rules allow them.",
    ));
    lines.extend([String::new(), "## Risky Actions".to_string(), String::new()]);
    lines.extend(list_or_default(
        &scan.risky_actions,
        "Ask before write, delete, email, payment, or admin work.",
    ));
    lines.extend([
        String::new(),
        "## Limits".to_string(),
        String::new(),
        format!("- {}", sentence(&scan.limits)),
        "- Stop after repeated errors.".to_string(),
        String::new(),
        "## More Info".to_string(),
        String::new(),
    ]);
    lines.extend(more_info(scan));
    lines.extend([
        String::new(),
        "## Credits".to_string(),
        String::new(),
        format!("- Spec: {SPEC_URL}"),
        format!("- Spec version: {SPEC_VERSION}"),
        format!("- Credit: {CREDIT_URL}"),
        "- License: CC-BY-4.0".to_string(),
    ]);

    format!(
        "{}\n",
        redact_secrets(&compact_blank_lines(&lines.join("\n"))).trim()
    )
}

fn source_text(scan: &ScanResult) -> String {
    let count = scan.routes.len();
    if count == 0 {
        return "No public HTTP routes found".to_string();
    }
    let languages = if scan.languages.is_empty() {
        "project files".to_string()
    } else {
        scan.languages.join(", ")
    };
    if count == 1 {
        format!("1 route from {languages}")
    } else {
        format!("{count} routes from {languages}")
    }
}

fn list_or_default(items: &[String], fallback: &str) -> Vec<String> {
    if items.is_empty() {
        return vec![format!("- {fallback}")];
    }
    items
        .iter()
        .filter(|item| !item.trim().is_empty())
        .take(MAX_ACTIONS)
        .map(|item| format!("- {}", sentence(item)))
        .collect()
}

fn more_info(scan: &ScanResult) -> Vec<String> {
    let mut lines = Vec::new();
    for doc in scan.docs.iter().take(6) {
        lines.push(format!("- Docs: {doc}"));
    }
    if !scan.support.is_empty() {
        lines.push(format!("- Support: {}", scan.support));
    }
    for note in scan.notes.iter().take(4) {
        lines.push(format!("- Note: {}", sentence(note)));
    }
    if lines.is_empty() {
        lines.push("- Add docs in `autoskill.config.json`.".to_string());
    }
    lines
}

fn sentence(value: &str) -> String {
    let clean = value.split_whitespace().collect::<Vec<_>>().join(" ");
    if clean.is_empty() {
        return clean;
    }
    if clean.ends_with(['.', '!', '?', '`']) {
        clean
    } else {
        format!("{clean}.")
    }
}

fn empty_default<'a>(value: &'a str, fallback: &'a str) -> &'a str {
    if value.is_empty() {
        fallback
    } else {
        value
    }
}

fn compact_blank_lines(text: &str) -> String {
    let mut output = Vec::new();
    let mut last_blank = false;
    for line in text.lines() {
        let blank = line.trim().is_empty();
        if blank && last_blank {
            continue;
        }
        output.push(line);
        last_blank = blank;
    }
    output.join("\n")
}
